Library/Homebrew/language/haskell.rb

```diff
@@ -92,5 +92,50 @@ def install_cabal_package(*args, **options)
         end
       end
     end
+
+    module CabalV2
+      def cabal_sandbox(options = {})
+        pwd = Pathname.pwd
+        home = options[:home] || pwd
+
+        # pretend HOME is elsewhere, so that ~/.cabal is kept as untouched
+        # as possible (except for ~/.cabal/setup-exe-cache)
+        # https://github.com/haskell/cabal/issues/1234
+        saved_home = ENV["HOME"]
+        ENV["HOME"] = home
+
+        system "cabal", "v2-update"
+
+        yield
+
+        # restore the environment
+        ENV["HOME"] = saved_home
+      end
+
+      def cabal_install(*args)
+        # cabal hardcodes 64 as the maximum number of parallel jobs
+        # https://github.com/Homebrew/legacy-homebrew/issues/49509
+        make_jobs = (ENV.make_jobs > 64) ? 64 : ENV.make_jobs
+
+        # cabal-install's dependency-resolution backtracking strategy can easily
+        # need more than the default 2,000 maximum number of "backjumps," since
+        # Hackage is a fast-moving, rolling-release target. The highest known
+        # needed value by a formula at this time (February 2016) was 43,478 for
+        # git-annex, so 100,000 should be enough to avoid most gratuitous
+        # backjumps build failures.
+        system "cabal", "v2-install", "--jobs=#{make_jobs}", "--max-backjumps=100000", *args
+      end
+
+      def install_cabal_package(*args)
+        cabal_sandbox do
+          # install the program in the destination dir
+          cabal_install "--install-method=copy",
```

Comment:
Similarly, this could be another `std_cabal_v2_something`

Reply:
Would it be acceptable to ditch the `cabal_sandbox` stuff entirely, and just provide a `install_cabal_package` that Does The Right Thing?  Looking at all the Haskell packages in Homebrew, the only one that actually *needs* `cabal_sandbox` is [Agda](https://formulae.brew.sh/formula/agda), because it does some code generation using self-built tools.  And even that might be achievable in some other way; I didn't study it closely.

Reply:
I think it'd be preferable to move back to `system` calls and provide helpers that make it easy to do the right thing.

Reply:
These are the literal boilerplate `system` calls that I think would work for the vast majority of Haskell packages:

```
ENV["CABAL_DIR"] = Pathname.pwd/".cabal"
system "cabal", "v2-update"
system "cabal", "v2-install", "--jobs=#{ENV.make_jobs}", "--max-backjumps=100000", "--install-method=copy", "--installdir=#{bin}"
```

If we really want to simplify everything down to the simplest possible, then I guess we'd just define `std_cabal_v2_install_flags=["--jobs=#{ENV.make_jobs}", "--max-backjumps=100000", "--install-method=copy", "--installdir=#{bin}"]` (where?) and write most of the formulae as

```
ENV["CABAL_DIR"] = Pathname.pwd/".cabal"
system "cabal", "v2-update"
system "cabal", "v2-install", *std_cabal_v2_install_flags
```

(Apologies if there's any syntax errors - I'm not fluent in Ruby, and I didn't run this.)

Reply:
> ENV["CABAL_DIR"] = Pathname.pwd/".cabal"

Excuse the ignorance: what goes here? Would it be useful to cache them?



> write most of the formulae as

This seems pretty good!

Reply:
Normally `cabal` will read and write various things to `$HOME/.cabal`, including caches and configuration files.  I think the Homebrew principle is to not touch the user's `$HOME`, so setting `$CABAL_DIR` makes it look for all that stuff elsewhere (in the temporary build directory).

It *would* be possible to have a shared CABAL_DIR owned by Homebrew, and used for all Haskell packages.  I'm not convinced it's a good idea, because it would only be effective for packages that use the *exact* same dependencies (including their transitive dependencies), and it would be a potentially fragile moving part.

Reply:
In an ideal situation we'd figure out a way to just cache the stuff that makes sense but agreed that otherwise it's just worth sticking somewhere in `buildpath`. That's done by default, though: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/formula.rb#L2065-L2073 (and we could add an extra variable there in general if it's needed).

Reply:
Am I understanding it correctly that Homebrew already sets `$HOME` to a temporary location while building formulae?  That would be good news, because then we can cut off the `ENV` part from my boilerplate.

Reply:
> Am I understanding it correctly that Homebrew already sets `$HOME` to a temporary location while building formulae?

Yup! I suspect the `cabal` `HOME` stuff predates that change.